pydantic_ai_slim/pydantic_ai/agent.py

```diff
@@ -1058,3 +1157,269 @@ def _prepare_result_schema(
+
+        agent = Agent('openai:gpt-4o')
+
+        async def main():
+            nodes = []
+            with agent.iter('What is the capital of France?') as agent_run:
+                # The first node can be retrieved via __anext__(), or you might already have it.
+                next_node = await agent_run.__anext__()
+                while not isinstance(next_node, End):
+                    next_node = await agent_run.next(next_node)
+                    nodes.append(next_node)
+                # Once `next_node` is an End, we've finished:
+                print(nodes)
+                '''
+                [
+                    HandleResponseNode(
+                        model_response=ModelResponse(
+                            parts=[TextPart(content='Paris', part_kind='text')],
+                            model_name='function:model_logic',
+                            timestamp=datetime.datetime(...),
+                            kind='response',
+                        )
+                    ),
+                    End(data=FinalResult(data='Paris', tool_name=None)),
+                ]
+                '''
+                print('Final result:', agent_run.final_result.data)
+                #> Final result: Paris
+        ```
+
+        Args:
+            node: The node to run next in the graph.
+
+        Returns:
+            The next node returned by the graph logic, or an [`End`][pydantic_graph.nodes.End] node if
+            the run has completed.
+        """
+        # Note: It might be nice to expose a synchronous interface for iteration, but we shouldn't do it
+        # on this class or IDEs won't warn you if you accidentally use `for` instead of `async for` to iterate.
+        return await self._graph_run.next(node)
+
+    def usage(self) -> _usage.Usage:
+        """Get usage statistics for the run so far, including token usage, model requests, and so on."""
+        return self._graph_run.state.usage
+
+    def __repr__(self):
+        final_result = self._graph_run.final_result
+        result_repr = '<run not finished>' if final_result is None else repr(final_result.result)
+        kws = [f'result={result_repr}', f'usage={self.usage()}']
+        return '<{} {}>'.format(type(self).__name__, ' '.join(kws))
+
+
+@dataclasses.dataclass
+class AgentRunResult(Generic[AgentDepsT, ResultDataT]):
+    """The final result of an agent run."""
+
+    graph_run_result: GraphRunResult[
+        _agent_graph.GraphAgentState, _agent_graph.GraphAgentDeps[AgentDepsT, Any], FinalResult[ResultDataT]
+    ]
+
+    @property
+    def _result(self) -> FinalResult[ResultDataT]:
+        return self.graph_run_result.result
+
+    @property
+    def data(self) -> ResultDataT:
+        return self._result.data
+
+    @property
+    def _result_tool_name(self) -> str | None:
+        return self._result.tool_name
+
+    def _set_result_tool_return(self, return_content: str) -> list[_messages.ModelMessage]:
```

Comment:
should this be public?

Reply:
🤷‍♂️ it isn't on main, and I'd rather not change that now; more generally I'd like to find a pattern that allows us to eliminate this in some way (or somehow make it less of a "mutating" API), though admittedly I haven't thought about it super deeply.